File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-eqd-5-4/EquityOption.cpp
// EquityOption.cpp 
#include "EquityOption.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

EquityOption::EquityOption(TiXmlNode* xmlNode)
: EquityDerivativeLongFormBase(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //strikeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* strikeNode = xmlNode->FirstChildElement("strike");

   if(strikeNode){strikeIsNull_ = false;}
   else{strikeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- strikeNode , address : " << strikeNode << std::endl;
   #endif
   if(strikeNode)
   {
      if(strikeNode->Attribute("href") || strikeNode->Attribute("id"))
      {
          if(strikeNode->Attribute("id"))
          {
             strikeIDRef_ = strikeNode->Attribute("id");
             strike_ = boost::shared_ptr<EquityStrike>(new EquityStrike(strikeNode));
             strike_->setID(strikeIDRef_);
             IDManager::instance().setID(strikeIDRef_,strike_);
          }
          else if(strikeNode->Attribute("href")) { strikeIDRef_ = strikeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { strike_ = boost::shared_ptr<EquityStrike>(new EquityStrike(strikeNode));}
   }

   //spotPriceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* spotPriceNode = xmlNode->FirstChildElement("spotPrice");

   if(spotPriceNode){spotPriceIsNull_ = false;}
   else{spotPriceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- spotPriceNode , address : " << spotPriceNode << std::endl;
   #endif
   if(spotPriceNode)
   {
      if(spotPriceNode->Attribute("href") || spotPriceNode->Attribute("id"))
      {
          if(spotPriceNode->Attribute("id"))
          {
             spotPriceIDRef_ = spotPriceNode->Attribute("id");
             spotPrice_ = boost::shared_ptr<NonNegativeDecimal>(new NonNegativeDecimal(spotPriceNode));
             spotPrice_->setID(spotPriceIDRef_);
             IDManager::instance().setID(spotPriceIDRef_,spotPrice_);
          }
          else if(spotPriceNode->Attribute("href")) { spotPriceIDRef_ = spotPriceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { spotPrice_ = boost::shared_ptr<NonNegativeDecimal>(new NonNegativeDecimal(spotPriceNode));}
   }

   //numberOfOptionsNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* numberOfOptionsNode = xmlNode->FirstChildElement("numberOfOptions");

   if(numberOfOptionsNode){numberOfOptionsIsNull_ = false;}
   else{numberOfOptionsIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- numberOfOptionsNode , address : " << numberOfOptionsNode << std::endl;
   #endif
   if(numberOfOptionsNode)
   {
      if(numberOfOptionsNode->Attribute("href") || numberOfOptionsNode->Attribute("id"))
      {
          if(numberOfOptionsNode->Attribute("id"))
          {
             numberOfOptionsIDRef_ = numberOfOptionsNode->Attribute("id");
             numberOfOptions_ = boost::shared_ptr<NonNegativeDecimal>(new NonNegativeDecimal(numberOfOptionsNode));
             numberOfOptions_->setID(numberOfOptionsIDRef_);
             IDManager::instance().setID(numberOfOptionsIDRef_,numberOfOptions_);
          }
          else if(numberOfOptionsNode->Attribute("href")) { numberOfOptionsIDRef_ = numberOfOptionsNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { numberOfOptions_ = boost::shared_ptr<NonNegativeDecimal>(new NonNegativeDecimal(numberOfOptionsNode));}
   }

   //optionEntitlementNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* optionEntitlementNode = xmlNode->FirstChildElement("optionEntitlement");

   if(optionEntitlementNode){optionEntitlementIsNull_ = false;}
   else{optionEntitlementIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- optionEntitlementNode , address : " << optionEntitlementNode << std::endl;
   #endif
   if(optionEntitlementNode)
   {
      if(optionEntitlementNode->Attribute("href") || optionEntitlementNode->Attribute("id"))
      {
          if(optionEntitlementNode->Attribute("id"))
          {
             optionEntitlementIDRef_ = optionEntitlementNode->Attribute("id");
             optionEntitlement_ = boost::shared_ptr<PositiveDecimal>(new PositiveDecimal(optionEntitlementNode));
             optionEntitlement_->setID(optionEntitlementIDRef_);
             IDManager::instance().setID(optionEntitlementIDRef_,optionEntitlement_);
          }
          else if(optionEntitlementNode->Attribute("href")) { optionEntitlementIDRef_ = optionEntitlementNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { optionEntitlement_ = boost::shared_ptr<PositiveDecimal>(new PositiveDecimal(optionEntitlementNode));}
   }

   //equityPremiumNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* equityPremiumNode = xmlNode->FirstChildElement("equityPremium");

   if(equityPremiumNode){equityPremiumIsNull_ = false;}
   else{equityPremiumIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- equityPremiumNode , address : " << equityPremiumNode << std::endl;
   #endif
   if(equityPremiumNode)
   {
      if(equityPremiumNode->Attribute("href") || equityPremiumNode->Attribute("id"))
      {
          if(equityPremiumNode->Attribute("id"))
          {
             equityPremiumIDRef_ = equityPremiumNode->Attribute("id");
             equityPremium_ = boost::shared_ptr<EquityPremium>(new EquityPremium(equityPremiumNode));
             equityPremium_->setID(equityPremiumIDRef_);
             IDManager::instance().setID(equityPremiumIDRef_,equityPremium_);
          }
          else if(equityPremiumNode->Attribute("href")) { equityPremiumIDRef_ = equityPremiumNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { equityPremium_ = boost::shared_ptr<EquityPremium>(new EquityPremium(equityPremiumNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<EquityStrike> EquityOption::getStrike()
{
   if(!this->strikeIsNull_){
        if(strikeIDRef_ != ""){
             return boost::shared_static_cast<EquityStrike>(IDManager::instance().getID(strikeIDRef_));
        }else{
             return this->strike_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<EquityStrike>();
   }
}
boost::shared_ptr<NonNegativeDecimal> EquityOption::getSpotPrice()
{
   if(!this->spotPriceIsNull_){
        if(spotPriceIDRef_ != ""){
             return boost::shared_static_cast<NonNegativeDecimal>(IDManager::instance().getID(spotPriceIDRef_));
        }else{
             return this->spotPrice_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<NonNegativeDecimal>();
   }
}
boost::shared_ptr<NonNegativeDecimal> EquityOption::getNumberOfOptions()
{
   if(!this->numberOfOptionsIsNull_){
        if(numberOfOptionsIDRef_ != ""){
             return boost::shared_static_cast<NonNegativeDecimal>(IDManager::instance().getID(numberOfOptionsIDRef_));
        }else{
             return this->numberOfOptions_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<NonNegativeDecimal>();
   }
}
boost::shared_ptr<PositiveDecimal> EquityOption::getOptionEntitlement()
{
   if(!this->optionEntitlementIsNull_){
        if(optionEntitlementIDRef_ != ""){
             return boost::shared_static_cast<PositiveDecimal>(IDManager::instance().getID(optionEntitlementIDRef_));
        }else{
             return this->optionEntitlement_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PositiveDecimal>();
   }
}
boost::shared_ptr<EquityPremium> EquityOption::getEquityPremium()
{
   if(!this->equityPremiumIsNull_){
        if(equityPremiumIDRef_ != ""){
             return boost::shared_static_cast<EquityPremium>(IDManager::instance().getID(equityPremiumIDRef_));
        }else{
             return this->equityPremium_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<EquityPremium>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-eqd-5-4/EquityOption.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class EquityOption : EquityDerivativeLongFormBase
    {
        public EquityOption(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode strikeNode = xmlNode.SelectSingleNode("strike");
            
            if (strikeNode != null)
            {
                if (strikeNode.Attributes["href"] != null || strikeNode.Attributes["id"] != null) 
                {
                    if (strikeNode.Attributes["id"] != null) 
                    {
                        strikeIDRef_ = strikeNode.Attributes["id"].Value;
                        EquityStrike ob = new EquityStrike(strikeNode);
                        IDManager.SetID(strikeIDRef_, ob);
                    }
                    else if (strikeNode.Attributes["href"] != null)
                    {
                        strikeIDRef_ = strikeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        strike_ = new EquityStrike(strikeNode);
                    }
                }
                else
                {
                    strike_ = new EquityStrike(strikeNode);
                }
            }
            
        
            XmlNode spotPriceNode = xmlNode.SelectSingleNode("spotPrice");
            
            if (spotPriceNode != null)
            {
                if (spotPriceNode.Attributes["href"] != null || spotPriceNode.Attributes["id"] != null) 
                {
                    if (spotPriceNode.Attributes["id"] != null) 
                    {
                        spotPriceIDRef_ = spotPriceNode.Attributes["id"].Value;
                        NonNegativeDecimal ob = new NonNegativeDecimal(spotPriceNode);
                        IDManager.SetID(spotPriceIDRef_, ob);
                    }
                    else if (spotPriceNode.Attributes["href"] != null)
                    {
                        spotPriceIDRef_ = spotPriceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        spotPrice_ = new NonNegativeDecimal(spotPriceNode);
                    }
                }
                else
                {
                    spotPrice_ = new NonNegativeDecimal(spotPriceNode);
                }
            }
            
        
            XmlNode numberOfOptionsNode = xmlNode.SelectSingleNode("numberOfOptions");
            
            if (numberOfOptionsNode != null)
            {
                if (numberOfOptionsNode.Attributes["href"] != null || numberOfOptionsNode.Attributes["id"] != null) 
                {
                    if (numberOfOptionsNode.Attributes["id"] != null) 
                    {
                        numberOfOptionsIDRef_ = numberOfOptionsNode.Attributes["id"].Value;
                        NonNegativeDecimal ob = new NonNegativeDecimal(numberOfOptionsNode);
                        IDManager.SetID(numberOfOptionsIDRef_, ob);
                    }
                    else if (numberOfOptionsNode.Attributes["href"] != null)
                    {
                        numberOfOptionsIDRef_ = numberOfOptionsNode.Attributes["href"].Value;
                    }
                    else
                    {
                        numberOfOptions_ = new NonNegativeDecimal(numberOfOptionsNode);
                    }
                }
                else
                {
                    numberOfOptions_ = new NonNegativeDecimal(numberOfOptionsNode);
                }
            }
            
        
            XmlNode optionEntitlementNode = xmlNode.SelectSingleNode("optionEntitlement");
            
            if (optionEntitlementNode != null)
            {
                if (optionEntitlementNode.Attributes["href"] != null || optionEntitlementNode.Attributes["id"] != null) 
                {
                    if (optionEntitlementNode.Attributes["id"] != null) 
                    {
                        optionEntitlementIDRef_ = optionEntitlementNode.Attributes["id"].Value;
                        PositiveDecimal ob = new PositiveDecimal(optionEntitlementNode);
                        IDManager.SetID(optionEntitlementIDRef_, ob);
                    }
                    else if (optionEntitlementNode.Attributes["href"] != null)
                    {
                        optionEntitlementIDRef_ = optionEntitlementNode.Attributes["href"].Value;
                    }
                    else
                    {
                        optionEntitlement_ = new PositiveDecimal(optionEntitlementNode);
                    }
                }
                else
                {
                    optionEntitlement_ = new PositiveDecimal(optionEntitlementNode);
                }
            }
            
        
            XmlNode equityPremiumNode = xmlNode.SelectSingleNode("equityPremium");
            
            if (equityPremiumNode != null)
            {
                if (equityPremiumNode.Attributes["href"] != null || equityPremiumNode.Attributes["id"] != null) 
                {
                    if (equityPremiumNode.Attributes["id"] != null) 
                    {
                        equityPremiumIDRef_ = equityPremiumNode.Attributes["id"].Value;
                        EquityPremium ob = new EquityPremium(equityPremiumNode);
                        IDManager.SetID(equityPremiumIDRef_, ob);
                    }
                    else if (equityPremiumNode.Attributes["href"] != null)
                    {
                        equityPremiumIDRef_ = equityPremiumNode.Attributes["href"].Value;
                    }
                    else
                    {
                        equityPremium_ = new EquityPremium(equityPremiumNode);
                    }
                }
                else
                {
                    equityPremium_ = new EquityPremium(equityPremiumNode);
                }
            }
            
        
        }
        
    
        #region Strike_
        private EquityStrike strike_;
        public EquityStrike Strike_
        {
            get
            {
                if (this.strike_ != null)
                {
                    return this.strike_; 
                }
                else if (this.strikeIDRef_ != null)
                {
                    strike_ = IDManager.getID(strikeIDRef_) as EquityStrike;
                    return this.strike_; 
                }
                else
                {
                      return this.strike_; 
                }
            }
            set
            {
                if (this.strike_ != value)
                {
                    this.strike_ = value;
                }
            }
        }
        #endregion
        
        public string strikeIDRef_ { get; set; }
        #region SpotPrice_
        private NonNegativeDecimal spotPrice_;
        public NonNegativeDecimal SpotPrice_
        {
            get
            {
                if (this.spotPrice_ != null)
                {
                    return this.spotPrice_; 
                }
                else if (this.spotPriceIDRef_ != null)
                {
                    spotPrice_ = IDManager.getID(spotPriceIDRef_) as NonNegativeDecimal;
                    return this.spotPrice_; 
                }
                else
                {
                      return this.spotPrice_; 
                }
            }
            set
            {
                if (this.spotPrice_ != value)
                {
                    this.spotPrice_ = value;
                }
            }
        }
        #endregion
        
        public string spotPriceIDRef_ { get; set; }
        #region NumberOfOptions_
        private NonNegativeDecimal numberOfOptions_;
        public NonNegativeDecimal NumberOfOptions_
        {
            get
            {
                if (this.numberOfOptions_ != null)
                {
                    return this.numberOfOptions_; 
                }
                else if (this.numberOfOptionsIDRef_ != null)
                {
                    numberOfOptions_ = IDManager.getID(numberOfOptionsIDRef_) as NonNegativeDecimal;
                    return this.numberOfOptions_; 
                }
                else
                {
                      return this.numberOfOptions_; 
                }
            }
            set
            {
                if (this.numberOfOptions_ != value)
                {
                    this.numberOfOptions_ = value;
                }
            }
        }
        #endregion
        
        public string numberOfOptionsIDRef_ { get; set; }
        #region OptionEntitlement_
        private PositiveDecimal optionEntitlement_;
        public PositiveDecimal OptionEntitlement_
        {
            get
            {
                if (this.optionEntitlement_ != null)
                {
                    return this.optionEntitlement_; 
                }
                else if (this.optionEntitlementIDRef_ != null)
                {
                    optionEntitlement_ = IDManager.getID(optionEntitlementIDRef_) as PositiveDecimal;
                    return this.optionEntitlement_; 
                }
                else
                {
                      return this.optionEntitlement_; 
                }
            }
            set
            {
                if (this.optionEntitlement_ != value)
                {
                    this.optionEntitlement_ = value;
                }
            }
        }
        #endregion
        
        public string optionEntitlementIDRef_ { get; set; }
        #region EquityPremium_
        private EquityPremium equityPremium_;
        public EquityPremium EquityPremium_
        {
            get
            {
                if (this.equityPremium_ != null)
                {
                    return this.equityPremium_; 
                }
                else if (this.equityPremiumIDRef_ != null)
                {
                    equityPremium_ = IDManager.getID(equityPremiumIDRef_) as EquityPremium;
                    return this.equityPremium_; 
                }
                else
                {
                      return this.equityPremium_; 
                }
            }
            set
            {
                if (this.equityPremium_ != value)
                {
                    this.equityPremium_ = value;
                }
            }
        }
        #endregion
        
        public string equityPremiumIDRef_ { get; set; }
        
    
        
    
    }
    
}

